arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+          other->_queue.pop_back();
+          if (toSteal > 1) {
+            std::unique_lock guard2(mutex);
+            for (std::size_t i = 1; i < toSteal; ++i) {
+              _queue.push_front(std::move(work));
+
+              TRI_ASSERT(other->_queue.size() > 0);
+              work = std::move(other->_queue.back());
+              TRI_ASSERT(work != nullptr);
+              other->_queue.pop_back();
+            }
+            queueSize.store(_queue.size(), std::memory_order_relaxed);
+          }
+          other->queueSize.store(other->_queue.size(),
+                                 std::memory_order_relaxed);
+
+          // remember this thread so next time we try to steal from it first
+          stealHint.store(idx, std::memory_order_relaxed);
+          return work;
+        }
+      }
+    }
+    if (this->pushQueue.load(std::memory_order_relaxed) != nullptr) {
+      // work has been pushed into our queue -> no need to still anything
+      return {};
+    }
+    idx += stepSize;
+    if (idx >= pool.numThreads) {
+      idx -= pool.numThreads;
+    }
+  }
+  // nothing found - reset the hint so next time we start at a random index
+  stealHint.store(NoHint, std::memory_order_relaxed);
+  return {};
+}
+
+WorkStealingThreadPool::WorkStealingThreadPool(const char* name,
+                                               std::size_t threadCount,
+                                               ThreadPoolMetrics metrics)
+    : numThreads(threadCount),
+      _metrics(metrics),
+      _threads(),
+      _threadStates(threadCount),
+      _latch(threadCount) {
+  auto coprimes = calculateCoprimes(threadCount);
+  std::size_t coprimeIdx = 0;
+  for (std::size_t i = 0; i < threadCount; ++i) {
+    _threads.emplace_back(std::jthread([this, i, &coprimes, coprimeIdx]() {
+      // we intentionally create the ThreadStates _inside_ the thread
+      _threadStates[i] =
+          std::make_unique<ThreadState>(i, coprimes[coprimeIdx], *this);
+
+      _threadStates[i]->run();
+    }));
+    ++coprimeIdx;
+    if (coprimeIdx >= coprimes.size()) {
+      coprimeIdx = 0;
+    }
+
+#ifdef TRI_HAVE_SYS_PRCTL_H
+    pthread_setname_np(_threads.back().native_handle(), name);
+#endif
+  }
+  // wait until all threads are initialized
+  _latch.wait();
+}
+
+WorkStealingThreadPool::~WorkStealingThreadPool() { shutdown(); }
+
+void WorkStealingThreadPool::shutdown() noexcept {
+  for (auto& thread : _threadStates) {
+    thread->signalStop();
+  }
+
+  for (auto& thread : _threads) {
+    if (thread.joinable()) {
+      thread.join();
+    }
+  }
+}
+
+void WorkStealingThreadPool::push(std::unique_ptr<WorkItem>&& task) noexcept {
+  if (auto* ts = ThreadState::currentThread;
+      ts != nullptr && ts->belongsToPool(*this)) {
+    ts->pushFront(std::move(task));
+  } else {
+    auto h = hint.load(std::memory_order_relaxed);
+    if (h != NoHint) {
+      auto prevHint =
+          _threadStates[h]->prevHint.load(std::memory_order_relaxed);
```

Comment:
Can there be a situation in which `hint` is set to anything else but `NoHint`, but `prevHint` still is `NoHint`?
In this case this would cause an out-of-bounds read, but I am not sure if it is guaranteed that `prevHint` always has a valid value when we get here.

Reply:
> Can there be a situation in which hint is set to anything else but `NoHint`, but `prevHint` still is `NoHint`?
Sure, but I don't see how this can cause any issues? `hint` and `prevHint` must always either contain a valid index or `NoHint`. If `hint != NoHint` but `_threadStates[h]->prevHint == NoHint` we simply update `hint` to `NoHint`.

Reply:
Ah true, please ignore my comment then.